In [62]:
import numpy as np
import pandas as pd
import scipy.stats as ss


In [2]:
closest_collection = "typeIII_submission_collection_closest.csv"
hungarian_collection = "typeIII_submission_collection_hungarian.csv"

## How many predicted pKas are matched differently between closest and hungarian algorithms?

In [3]:
df_closest = pd.read_csv(closest_collection,index_col=0)
df_closest.head()

$\Delta$pKa error (calc - exp) Molecule ID               name  pKa (calc)  \
0                            0.18        SM01  COSMOtherm_FINE17        9.71   
1                            1.18        SM02  COSMOtherm_FINE17        6.21   
2                           -6.39        SM03  COSMOtherm_FINE17        0.63   
3                           -0.61        SM04  COSMOtherm_FINE17        5.41   
4                           -0.06        SM05  COSMOtherm_FINE17        4.53   

   pKa (exp)     pKa ID  pKa SEM (calc)  pKa SEM (exp)    participant  \
0       9.53  SM01_pKa1            0.65           0.01  Chris Loschen   
1       5.03  SM02_pKa1            0.65           0.01  Chris Loschen   
2       7.02  SM03_pKa1            0.65           0.01  Chris Loschen   
3       6.02  SM04_pKa1            0.65           0.01  Chris Loschen   
4       4.59  SM05_pKa1            0.65           0.01  Chris Loschen   

  receipt_id  
0      0hxtm  
1      0hxtm  
2      0hxtm  
3      0hxtm  
4      0hxtm

In [34]:
df_hungarian = pd.read_csv(hungarian_collection, index_col=0)
df_hungarian.head()

$\Delta$pKa error (calc - exp) Molecule ID               name  pKa (calc)  \
0                            0.18        SM01  COSMOtherm_FINE17        9.71   
1                            1.18        SM02  COSMOtherm_FINE17        6.21   
2                           -6.39        SM03  COSMOtherm_FINE17        0.63   
3                           -0.61        SM04  COSMOtherm_FINE17        5.41   
4                           -0.06        SM05  COSMOtherm_FINE17        4.53   

   pKa (exp)     pKa ID  pKa SEM (calc)  pKa SEM (exp)    participant  \
0       9.53  SM01_pKa1            0.65           0.01  Chris Loschen   
1       5.03  SM02_pKa1            0.65           0.01  Chris Loschen   
2       7.02  SM03_pKa1            0.65           0.01  Chris Loschen   
3       6.02  SM04_pKa1            0.65           0.01  Chris Loschen   
4       4.59  SM05_pKa1            0.65           0.01  Chris Loschen   

  receipt_id  
0      0hxtm  
1      0hxtm  
2      0hxtm  
3      0hxtm  
4      0hxtm

In [5]:
prediction_methods = set(df_closest["name"])
len(prediction_methods)

34

In [28]:
# Iterate through prediction methods and create a database that compares hunagarian and closest matching

matched_pKa_list = []

for method in prediction_methods:
#for method in ["Full quantum chemical calculation of free energies and fit to experimental pKa"]:
    
    submission_id = df_closest[df_closest["name"] == method]["receipt_id"].values[0]
    
    df_closest_1method = df_closest[df_closest["name"] == method].reset_index(drop=True)    
    df_hungarian_1method = df_hungarian[df_hungarian["name"] == method].reset_index(drop=True)
     
    pKa_IDs = list(df_closest_1method["pKa ID"])
  
    # Iterate through pKa_IDs to check if predicted pKas match
    for pKa_ID in pKa_IDs:
        pKa_exp = df_closest_1method[df_closest_1method["pKa ID"] == pKa_ID]["pKa (exp)"].values[0]
        pKa_pred_closest = df_closest_1method[df_closest_1method["pKa ID"] == pKa_ID]["pKa (calc)"].values[0]
        pKa_pred_hungarian = df_hungarian_1method[df_hungarian_1method["pKa ID"] == pKa_ID]["pKa (calc)"].values[0]
        closest_hungarian_diff = pKa_pred_closest - pKa_pred_hungarian
                
        matched_pKa_row = [pKa_ID, pKa_exp, pKa_pred_closest, pKa_pred_hungarian, closest_hungarian_diff, submission_id]
        matched_pKa_list.append(matched_pKa_row)
        #print(matched_pKa_row)

# Convert to pandas dataframe
df_compare_matching = pd.DataFrame(matched_pKa_list, columns = ['pKa ID', 'pKa (exp)', 
                                               'pKa (pred, closest)', 'pKa (pred, hungarian)',
                                              'closest - hungarian diff.', 'submission ID'])
df_compare_matching

pKa ID  pKa (exp)  pKa (pred, closest)  pKa (pred, hungarian)  \
0    SM01_pKa1       9.53                 9.27                   9.27   
1    SM02_pKa1       5.03                 5.19                   5.19   
2    SM03_pKa1       7.02                 4.49                   4.49   
3    SM04_pKa1       6.02                 6.73                   6.73   
4    SM05_pKa1       4.59                 7.62                   7.62   
5    SM06_pKa1       3.03                 3.94                   1.77   
6    SM07_pKa1       6.08                 5.17                   5.17   
7    SM08_pKa1       4.22                 4.61                   4.61   
8    SM09_pKa1       5.37                 5.14                   5.14   
9    SM10_pKa1       9.02                 6.44                   6.44   
10   SM11_pKa1       3.89                 5.07                   5.07   
11   SM12_pKa1       5.28                 5.17                   5.17   
12   SM13_pKa1       5.77                 4.97                   4.97   
13   SM14_pKa1       2.58                 0.12                   0.12   
14   SM14_pKa2       5.30                 6.49                   6.49   
15   SM15_pKa1       4.70                 5.42                   5.42   
16   SM15_pKa2       8.94                 8.71                   8.71   
17   SM16_pKa1       5.37                 5.91                   5.91   
18   SM17_pKa1       3.16                 3.47                   3.47   
19   SM18_pKa1       2.15                -0.26                  -0.26   
20   SM18_pKa3      11.02                10.98                  10.98   
21   SM19_pKa1       9.56                 6.04                   6.04   
22   SM20_pKa1       5.70                 7.31                   7.31   
23   SM21_pKa1       4.10                 4.07                   4.07   
24   SM22_pKa1       2.40                 2.73                   2.73   
25   SM22_pKa2       7.43                 6.60                   6.60   
26   SM23_pKa1       5.45                 5.48                   5.48   
27   SM24_pKa1       2.60                 1.71                   1.71   
28   SM01_pKa1       9.53                 9.12                   9.12   
29   SM02_pKa1       5.03                 4.05                   4.05   
..         ...        ...                  ...                    ...   
927  SM01_pKa1       9.53                11.36                  11.36   
928  SM02_pKa1       5.03                 3.95                   3.95   
929  SM03_pKa1       7.02                 6.23                   6.23   
930  SM04_pKa1       6.02                 5.23                   5.23   
931  SM05_pKa1       4.59                 2.11                   2.11   
932  SM06_pKa2      11.74                13.30                  13.30   
933  SM06_pKa1       3.03                -0.67                  -0.67   
934  SM07_pKa1       6.08                 6.02                   6.02   
935  SM08_pKa1       4.22                 3.17                   3.17   
936  SM09_pKa1       5.37                 4.14                   4.14   
937  SM10_pKa1       9.02                 8.36                   8.36   
938  SM11_pKa1       3.89                 1.43                   1.43   
939  SM12_pKa1       5.28                 3.57                   3.57   
940  SM13_pKa1       5.77                 3.52                   3.52   
941  SM14_pKa1       2.58                 3.64                   3.64   
942  SM15_pKa2       8.94                 9.85                   9.85   
943  SM15_pKa1       4.70                 2.57                   2.57   
944  SM16_pKa2      10.65                11.18                  11.18   
945  SM16_pKa1       5.37                 2.43                   2.43   
946  SM17_pKa1       3.16                 0.21                   0.21   
947  SM18_pKa3      11.02                12.61                  12.61   
948  SM18_pKa2       9.58                 8.94                   8.94   
949  SM18_pKa1       2.15                 0.67                   0

In [29]:
# Print out pKas that have different matching between hungarian and closest
df_difference_in_matching = df_compare_matching[df_compare_matching["closest - hungarian diff."] != 0]
df_difference_in_matching

pKa ID  pKa (exp)  pKa (pred, closest)  pKa (pred, hungarian)  \
5    SM06_pKa1       3.03                 3.94                   1.77   
79   SM18_pKa3      11.02                10.56                  13.45   
142  SM18_pKa2       9.58                 9.87                   3.35   
162  SM14_pKa1       2.58                 3.81                   0.42   
163  SM14_pKa2       5.30                18.84                   3.81   
231  SM18_pKa2       9.58                10.02                   8.39   
254  SM14_pKa1       2.58                 3.38                  -1.77   
255  SM14_pKa2       5.30                24.63                   3.38   
262  SM18_pKa3      11.02                11.53                  11.75   
284  SM14_pKa1       2.58                 3.84                   0.47   
285  SM14_pKa2       5.30                19.09                   3.84   
314  SM14_pKa1       2.58                 3.74                  -1.40   
315  SM14_pKa2       5.30                18.61                   3.74   
327  SM22_pKa1       2.40                 4.23                  -3.77   
365  SM06_pKa1       3.03                 4.40                   1.38   
378  SM18_pKa1       2.15                 1.84                   0.50   
501  SM22_pKa1       2.40                 4.82                  -1.34   
587  SM18_pKa2       9.58                 9.26                   7.53   
616  SM18_pKa2       9.58                 9.59                   8.38   
639  SM14_pKa1       2.58                 3.86                  -0.90   
640  SM14_pKa2       5.30                23.88                   3.86   
669  SM14_pKa1       2.58                 3.94                   0.14   
670  SM14_pKa2       5.30                19.65                   3.94   
706  SM18_pKa2       9.58                 9.51                   7.59   
736  SM18_pKa2       9.58                 7.43                   6.60   
766  SM18_pKa3      11.02                10.97                  13.98   
796  SM18_pKa2       9.58                 9.91                   9.24   
818  SM14_pKa1       2.58                 3.94                   0.15   
819  SM14_pKa2       5.30                19.58                   3.94   
875  SM06_pKa1       3.03                 3.98                   0.98   

     closest - hungarian diff. submission ID  
5                         2.17         hytjn  
79                       -2.89         xikp8  
142                       6.52         pwn3m  
162                       3.39         mpwiy  
163                      15.03         mpwiy  
231                       1.63         w4iyd  
254                       5.15         nb006  
255                      21.25         nb006  
262                      -0.22         nb006  
284                       3.37         ttjd0  
285                      15.25         ttjd0  
314                       5.14         ds62k  
315                      14.87         ds62k  
327                       8.00         ds62k  
365                       3.02         0hxtm  
378                       1.34         0hxtm  
501                       6.16         2ii2g  
587                       1.73         yqkga  
616                       1.21         5byn6  
639                       4.76         nb005  
640                      20.02         nb005  
669                       3.80         nb003  
670                      15.71         nb003  
706                       1.92         ryzue  
736                       0.83         yc70m  
766                      -3.01         ad5pu  
796                       0.67         nb002  
818                       3.79         nb004  
819                      15.64         nb004  
875                       3.00         q3pfp

In [32]:
# Why is nb006 SM14_pKa2 matched to a very different number?

# These are submitted SM14 predictions
# SM14, -1.77, 1.77
# SM14,  3.38, 1.77
# SM14, 24.63, 1.77


# SM14 Experimental values
# 2.58 ± 0.01
# 5.30 ± 0.01


df_nb006 = df_closest[df_closest['receipt_id']=='nb006']
df_nb006_SM14 = df_nb006[df_nb006['Molecule ID']=='SM14']
df_nb006_SM14

$\Delta$pKa error (calc - exp) Molecule ID  \
553                            0.80        SM14   
554                           19.33        SM14   

                                           name  pKa (calc)  pKa (exp)  \
553  EC-RISM/MP2/6-311+G(d,p)-P3NI-phi-all-1par        3.38       2.58   
554  EC-RISM/MP2/6-311+G(d,p)-P3NI-phi-all-1par       24.63       5.30   

        pKa ID  pKa SEM (calc)  pKa SEM (exp)  participant receipt_id  
553  SM14_pKa1            1.77           0.01  Stefan Kast      nb006  
554  SM14_pKa2            1.77           0.01  Stefan Kast      nb006

In [33]:
df_nb006 = df_hungarian[df_hungarian['receipt_id']=='nb006']
df_nb006_SM14 = df_nb006[df_nb006['Molecule ID']=='SM14']
df_nb006_SM14

$\Delta$pKa error (calc - exp) Molecule ID  \
568                           -4.35        SM14   
569                           -1.92        SM14   

                                           name  pKa (calc)  pKa (exp)  \
568  EC-RISM/MP2/6-311+G(d,p)-P3NI-phi-all-1par       -1.77       2.58   
569  EC-RISM/MP2/6-311+G(d,p)-P3NI-phi-all-1par        3.38       5.30   

        pKa ID  pKa SEM (calc)  pKa SEM (exp)  participant receipt_id  
568  SM14_pKa1            1.77           0.01  Stefan Kast      nb006  
569  SM14_pKa2            1.77           0.01  Stefan Kast      nb006

### Experimental pKas of molecules with pKas differently matched
SM06  
3.03 ± 0.04   
11.74 ± 0.01  

SM14  
2.58 ± 0.01   
5.30 ± 0.01  

SM18  
2.15 ± 0.02   
9.58 ± 0.03   
11.02 ± 0.04  

SM22  
2.40 ± 0.02  
7.43 ± 0.01  

### Experimental pKas of molecules with pKas equally matched even thought they have multiple pKas
SM15   
4.70 ± 0.01   
8.94 ± 0.01   

SM16  
5.37 ± 0.01   
10.65 ± 0.01  



## How many pKa predictions are matched without conserving the sequence with Hungarian method?

In [85]:
# Test for comparing rank orders - ORDERED MATCH
exp_pKas = np.array([2.4, 4.3, 7.0])
pred_pKas = np.array([2.5, 4.2, 7.2])

#exp_pKa_ranks = list(np.array([1, 3, 2]))
#pred_pKa_ranks = list(np.array([1, 3, 2]))

exp_pKa_ranks = ss.rankdata(exp_pKas)
print("exp ranks:", exp_pKa_ranks)
pred_pKa_ranks = ss.rankdata(pred_pKas)
print("pred ranks:", pred_pKa_ranks)

# Is rank order the same?
if list(exp_pKa_ranks) == list(pred_pKa_ranks):
    ordered_match = True
else:
    ordered_match = False
    
ordered_match

exp ranks: [ 1.  2.  3.]
pred ranks: [ 1.  2.  3.]


True

In [86]:
# Test for comparing rank orders - UNORDERED MATCH
exp_pKas = np.array([2.4, 4.3, 7.0])
pred_pKas = np.array([2.5, 7.0, 4.5])

#exp_pKa_ranks = list(np.array([1, 3, 2]))
#pred_pKa_ranks = list(np.array([1, 3, 2]))

exp_pKa_ranks = ss.rankdata(exp_pKas)
print("exp ranks:", exp_pKa_ranks)
pred_pKa_ranks = ss.rankdata(pred_pKas)
print("pred ranks:", pred_pKa_ranks)

# Is rank order the same?
if list(exp_pKa_ranks) == list(pred_pKa_ranks):
    ordered_match = True
else:
    ordered_match = False
    
ordered_match

exp ranks: [ 1.  2.  3.]
pred ranks: [ 1.  3.  2.]


False

In [35]:
df_hungarian.head()

$\Delta$pKa error (calc - exp) Molecule ID               name  pKa (calc)  \
0                            0.18        SM01  COSMOtherm_FINE17        9.71   
1                            1.18        SM02  COSMOtherm_FINE17        6.21   
2                           -6.39        SM03  COSMOtherm_FINE17        0.63   
3                           -0.61        SM04  COSMOtherm_FINE17        5.41   
4                           -0.06        SM05  COSMOtherm_FINE17        4.53   

   pKa (exp)     pKa ID  pKa SEM (calc)  pKa SEM (exp)    participant  \
0       9.53  SM01_pKa1            0.65           0.01  Chris Loschen   
1       5.03  SM02_pKa1            0.65           0.01  Chris Loschen   
2       7.02  SM03_pKa1            0.65           0.01  Chris Loschen   
3       6.02  SM04_pKa1            0.65           0.01  Chris Loschen   
4       4.59  SM05_pKa1            0.65           0.01  Chris Loschen   

  receipt_id  
0      0hxtm  
1      0hxtm  
2      0hxtm  
3      0hxtm  
4      0hxtm

In [87]:
pKa_rank_comparison_list =[]

# Iterate through methods
for method in prediction_methods:
#for method in ["Full quantum chemical calculation of free energies and fit to experimental pKa"]:

    submission_ID = df_hungarian[df_hungarian["name"] == method]["receipt_id"].values[0]
    df_hungarian_1method = df_hungarian[df_hungarian["name"] == method].reset_index(drop=True)
     
    mol_IDs = list(df_hungarian_1method["Molecule ID"])

    #Iterate through molecules
    for mol_ID in mol_IDs:
        df_hungarian_1method_1mol = df_hungarian_1method[df_hungarian_1method["Molecule ID"] == mol_ID].reset_index(drop=True)
        
        pKa_IDs = df_hungarian_1method_1mol['pKa ID'].values
        
        #Assign Rank order of experimental and predicted pKa.
        exp_pKas = df_hungarian_1method_1mol['pKa (exp)'].values 
        exp_pKa_ranks = ss.rankdata(exp_pKas )
        
        
        pred_pKas = df_hungarian_1method_1mol['pKa (calc)'].values
        pred_pKa_ranks = ss.rankdata(pred_pKas) # rank is given to only matched pred pKas

    
        # Is rank order the same?
        if list(exp_pKa_ranks) == list(pred_pKa_ranks):
            ordered_match = True
        else:
            ordered_match = False

        
        pKa_rank_comparison_list.append([mol_ID, ordered_match, pKa_IDs, exp_pKas, pred_pKas, exp_pKa_ranks, pred_pKa_ranks, submission_ID])
    
    
# Convert to pandas dataframe
df_compare_ranks_hungarian = pd.DataFrame(pKa_rank_comparison_list, columns = ['mol ID', 'ordered match', 'pKa IDs', 'pKa (exp)', 
                                               'pKa (pred)', 'pKa rank (exp)', 'pKa rank (pred)', 'submission ID'])
df_compare_ranks_hungarian

mol ID  ordered match                            pKa IDs  \
0     SM01           True                        [SM01_pKa1]   
1     SM02           True                        [SM02_pKa1]   
2     SM03           True                        [SM03_pKa1]   
3     SM04           True                        [SM04_pKa1]   
4     SM05           True                        [SM05_pKa1]   
5     SM06           True             [SM06_pKa1, SM06_pKa2]   
6     SM06           True             [SM06_pKa1, SM06_pKa2]   
7     SM07           True                        [SM07_pKa1]   
8     SM08           True                        [SM08_pKa1]   
9     SM09           True                        [SM09_pKa1]   
10    SM10           True                        [SM10_pKa1]   
11    SM11           True                        [SM11_pKa1]   
12    SM12           True                        [SM12_pKa1]   
13    SM13           True                        [SM13_pKa1]   
14    SM14           True             [SM14_pKa1, SM14_pKa2]   
15    SM14           True             [SM14_pKa1, SM14_pKa2]   
16    SM15           True             [SM15_pKa1, SM15_pKa2]   
17    SM15           True             [SM15_pKa1, SM15_pKa2]   
18    SM16           True                        [SM16_pKa1]   
19    SM17           True                        [SM17_pKa1]   
20    SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
21    SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
22    SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
23    SM19           True                        [SM19_pKa1]   
24    SM20           True                        [SM20_pKa1]   
25    SM21           True                        [SM21_pKa1]   
26    SM22           True             [SM22_pKa1, SM22_pKa2]   
27    SM22           True             [SM22_pKa1, SM22_pKa2]   
28    SM23           True                        [SM23_pKa1]   
29    SM24           True                        [SM24_pKa1]   
..     ...            ...                                ...   
955   SM01           True                        [SM01_pKa1]   
956   SM02           True                        [SM02_pKa1]   
957   SM03           True                        [SM03_pKa1]   
958   SM04           True                        [SM04_pKa1]   
959   SM05           True                        [SM05_pKa1]   
960   SM06           True             [SM06_pKa2, SM06_pKa1]   
961   SM06           True             [SM06_pKa2, SM06_pKa1]   
962   SM07           True                        [SM07_pKa1]   
963   SM08           True                        [SM08_pKa1]   
964   SM09           True                        [SM09_pKa1]   
965   SM10           True                        [SM10_pKa1]   
966   SM11           True                        [SM11_pKa1]   
967   SM12           True                        [SM12_pKa1]   
968   SM13           True                        [SM13_pKa1]   
969   SM14           True                        [SM14_pKa1]   
970   SM15           True             [SM15_pKa2, SM15_pKa1]   
971   SM15           True             [SM15_pKa2, SM15_pKa1]   
972   SM16           True             [SM16_pKa2, SM16_pKa1]   
973   SM16           True             [SM16_pKa2, SM16_pKa1]   
974   SM17           True                        [SM17_pKa1]   
975   SM18           True  [SM18_pKa3, SM18_pKa2, SM18_pKa1]   
976   SM18           True  [SM18_pKa3, SM18_pKa2, SM18_pKa1]   
977   SM18           True  [SM18_pKa3, SM18_pKa2, SM18_pKa1]   
978   SM19           True                        [SM19_pKa1]   
979   SM20           True                        [SM20_pKa1]   
980   SM21           True                        [SM21_pKa1]   
981   SM22           True             [SM22_pKa2, SM22_pKa1]   
982   SM22           True             [SM22_pKa2, SM22_pKa1]   
983   SM23           True                        [SM23_pKa1]   
984   SM24           True                        [SM24_pKa1]   

               pKa (exp)           

In [80]:
# Print out pKas that don't preserve increasing order when matched my Hungarian algorithm
df_unordered_matching_hungarian = df_compare_ranks_hungarian[df_compare_ranks_hungarian["ordered match"] == False]
df_unordered_matching_hungarian

Empty DataFrame
Columns: [mol ID, ordered match, pKa IDs, pKa (exp), pKa (pred), pKa rank (exp), pKa rank (pred), submission ID]
Index: []

In [89]:
# Just SM18
df_unordered_matching_hungarian_SM18 = df_compare_ranks_hungarian[df_compare_ranks_hungarian["mol ID"] == 'SM18']
df_unordered_matching_hungarian_SM18

mol ID  ordered match                            pKa IDs  \
20    SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
21    SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
22    SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
51    SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
52    SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
53    SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
81    SM18           True  [SM18_pKa3, SM18_pKa2, SM18_pKa1]   
82    SM18           True  [SM18_pKa3, SM18_pKa2, SM18_pKa1]   
83    SM18           True  [SM18_pKa3, SM18_pKa2, SM18_pKa1]   
111   SM18           True             [SM18_pKa1, SM18_pKa2]   
112   SM18           True             [SM18_pKa1, SM18_pKa2]   
145   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
146   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
147   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
174   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
175   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
176   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
205   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
206   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
207   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
236   SM18           True  [SM18_pKa3, SM18_pKa2, SM18_pKa1]   
237   SM18           True  [SM18_pKa3, SM18_pKa2, SM18_pKa1]   
238   SM18           True  [SM18_pKa3, SM18_pKa2, SM18_pKa1]   
267   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
268   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
269   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
298   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
299   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
300   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
329   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
..     ...            ...                                ...   
694   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
695   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
696   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
725   SM18           True  [SM18_pKa2, SM18_pKa3, SM18_pKa1]   
726   SM18           True  [SM18_pKa2, SM18_pKa3, SM18_pKa1]   
727   SM18           True  [SM18_pKa2, SM18_pKa3, SM18_pKa1]   
756   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
757   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
758   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
787   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
788   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
789   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
818   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
819   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
820   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
849   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
850   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
851   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
885   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
886   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
887   SM18           True  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
915   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
916   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
917   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
945   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
946   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
947   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
975   SM18           True  [SM18_pKa3, SM18_pKa2, SM18_pKa1]   
976   SM18           True  [SM18_pKa3, SM18_pKa2, SM18_pKa1]   
977   SM18           True  [SM18_pKa3, SM18_pKa2, SM18_pKa1]   

               pKa (exp)           

In [90]:
# SM18 prediction of 0hxtm method was also matched in an order.
df_unordered_matching_hungarian_SM18_0hxtm = df_unordered_matching_hungarian_SM18[df_compare_ranks_hungarian["submission ID"] == '0hxtm']
df_unordered_matching_hungarian_SM18_0hxtm

/Users/isikm/opt/anaconda/envs/sampl6_pKa/lib/python3.6/site-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


mol ID  ordered match                 pKa IDs     pKa (exp)   pKa (pred)  \
389   SM18           True  [SM18_pKa2, SM18_pKa1]  [9.58, 2.15]  [1.84, 0.5]   
390   SM18           True  [SM18_pKa2, SM18_pKa1]  [9.58, 2.15]  [1.84, 0.5]   

    pKa rank (exp) pKa rank (pred) submission ID  
389     [2.0, 1.0]      [2.0, 1.0]         0hxtm  
390     [2.0, 1.0]      [2.0, 1.0]         0hxtm

There isn't any matches out of order in this set.

### Was there a out of order match in the past for Hungarian matching? Is it random for SM18 in 0hxtm submission?

In [94]:
# SAMPL6 repository branch pKa_typeIII_analysis3_hungarian
# https://github.com/MobleyLab/SAMPL6/blob/pKa_typeIII_analysis3_hungarian/physical_properties/pKa/analysis/analysis_of_typeIII_predictions/analysis_outputs_hungarian/typeIII_submission_collection.csv
hungarian_collection_a3 = 'typeIII_submission_collection_hungarian_analysis3.csv'
df_hungarian = pd.read_csv(hungarian_collection_a3, index_col=0)


pKa_rank_comparison_list =[]

# Iterate through methods
for method in prediction_methods:
#for method in ["Full quantum chemical calculation of free energies and fit to experimental pKa"]:

    submission_ID = df_hungarian[df_hungarian["name"] == method]["receipt_id"].values[0]
    df_hungarian_1method = df_hungarian[df_hungarian["name"] == method].reset_index(drop=True)
     
    mol_IDs = list(df_hungarian_1method["Molecule ID"])

    #Iterate through molecules
    for mol_ID in mol_IDs:
        df_hungarian_1method_1mol = df_hungarian_1method[df_hungarian_1method["Molecule ID"] == mol_ID].reset_index(drop=True)
        
        pKa_IDs = df_hungarian_1method_1mol['pKa ID'].values
        
        #Assign Rank order of experimental and predicted pKa.
        exp_pKas = df_hungarian_1method_1mol['pKa (exp)'].values 
        exp_pKa_ranks = ss.rankdata(exp_pKas )
        
        
        pred_pKas = df_hungarian_1method_1mol['pKa (calc)'].values
        pred_pKa_ranks = ss.rankdata(pred_pKas) # rank is given to only matched pred pKas

    
        # Is rank order the same?
        if list(exp_pKa_ranks) == list(pred_pKa_ranks):
            ordered_match = True
        else:
            ordered_match = False

        
        pKa_rank_comparison_list.append([mol_ID, ordered_match, pKa_IDs, exp_pKas, pred_pKas, exp_pKa_ranks, pred_pKa_ranks, submission_ID])
    
    
# Convert to pandas dataframe
df_compare_ranks_hungarian = pd.DataFrame(pKa_rank_comparison_list, columns = ['mol ID', 'ordered match', 'pKa IDs', 'pKa (exp)', 
                                               'pKa (pred)', 'pKa rank (exp)', 'pKa rank (pred)', 'submission ID'])

# Print out pKas that don't preserve increasing order when matched my Hungarian algorithm
df_unordered_matching_hungarian = df_compare_ranks_hungarian[df_compare_ranks_hungarian["ordered match"] == False]
df_unordered_matching_hungarian

mol ID  ordered match                            pKa IDs  \
389   SM18          False             [SM18_pKa1, SM18_pKa2]   
390   SM18          False             [SM18_pKa1, SM18_pKa2]   
601   SM18          False  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
602   SM18          False  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
603   SM18          False  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
725   SM18          False  [SM18_pKa3, SM18_pKa2, SM18_pKa1]   
726   SM18          False  [SM18_pKa3, SM18_pKa2, SM18_pKa1]   
727   SM18          False  [SM18_pKa3, SM18_pKa2, SM18_pKa1]   
756   SM18          False  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
757   SM18          False  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   
758   SM18          False  [SM18_pKa1, SM18_pKa3, SM18_pKa2]   

               pKa (exp)          pKa (pred)   pKa rank (exp)  \
389         [2.15, 9.58]         [1.84, 0.5]       [1.0, 2.0]   
390         [2.15, 9.58]         [1.84, 0.5]       [1.0, 2.0]   
601  [2.15, 11.02, 9.58]  [2.14, 7.53, 9.26]  [1.0, 3.0, 2.0]   
602  [2.15, 11.02, 9.58]  [2.14, 7.53, 9.26]  [1.0, 3.0, 2.0]   
603  [2.15, 11.02, 9.58]  [2.14, 7.53, 9.26]  [1.0, 3.0, 2.0]   
725  [11.02, 9.58, 2.15]  [7.59, 9.51, 1.41]  [3.0, 2.0, 1.0]   
726  [11.02, 9.58, 2.15]  [7.59, 9.51, 1.41]  [3.0, 2.0, 1.0]   
727  [11.02, 9.58, 2.15]  [7.59, 9.51, 1.41]  [3.0, 2.0, 1.0]   
756  [2.15, 11.02, 9.58]   [4.29, 6.6, 7.43]  [1.0, 3.0, 2.0]   
757  [2.15, 11.02, 9.58]   [4.29, 6.6, 7.43]  [1.0, 3.0, 2.0]   
758  [2.15, 11.02, 9.58]   [4.29, 6.6, 7.43]  [1.0, 3.0, 2.0]   

     pKa rank (pred) submission ID  
389       [2.0, 1.0]         0hxtm  
390       [2.0, 1.0]         0hxtm  
601  [1.0, 2.0, 3.0]         yqkga  
602  [1.0, 2.0, 3.0]         yqkga  
603  [1.0, 2.0, 3.0]         yqkga  
725  [2.0, 3.0, 1.0]         ryzue  
726  [2.0, 3.0, 1.0]         ryzue  
727  [2.0, 3.0, 1.0]         ryzue  
756  [1.0, 2.0, 3.0]         yc70m  
757  [1.0, 2.0, 3.0]         yc70m  
758  [1.0, 2.0, 3.0]         yc70m

In [96]:
# SAMPL6 repository branch pKa_typeIII_analysis5_hungarian
# https://github.com/MobleyLab/SAMPL6/blob/pKa_typeIII_analysis5/physical_properties/pKa/analysis/analysis_of_typeIII_predictions/analysis_outputs_hungarian/typeIII_submission_collection.csv
hungarian_collection_a5 = 'typeIII_submission_collection_hungarian_analysis5.csv'
df_hungarian = pd.read_csv(hungarian_collection_a5, index_col=0)


pKa_rank_comparison_list =[]

# Iterate through methods
for method in prediction_methods:
#for method in ["Full quantum chemical calculation of free energies and fit to experimental pKa"]:

    submission_ID = df_hungarian[df_hungarian["name"] == method]["receipt_id"].values[0]
    df_hungarian_1method = df_hungarian[df_hungarian["name"] == method].reset_index(drop=True)
     
    mol_IDs = list(df_hungarian_1method["Molecule ID"])

    #Iterate through molecules
    for mol_ID in mol_IDs:
        df_hungarian_1method_1mol = df_hungarian_1method[df_hungarian_1method["Molecule ID"] == mol_ID].reset_index(drop=True)
        
        pKa_IDs = df_hungarian_1method_1mol['pKa ID'].values
        
        #Assign Rank order of experimental and predicted pKa.
        exp_pKas = df_hungarian_1method_1mol['pKa (exp)'].values 
        exp_pKa_ranks = ss.rankdata(exp_pKas )
        
        
        pred_pKas = df_hungarian_1method_1mol['pKa (calc)'].values
        pred_pKa_ranks = ss.rankdata(pred_pKas) # rank is given to only matched pred pKas

    
        # Is rank order the same?
        if list(exp_pKa_ranks) == list(pred_pKa_ranks):
            ordered_match = True
        else:
            ordered_match = False

        
        pKa_rank_comparison_list.append([mol_ID, ordered_match, pKa_IDs, exp_pKas, pred_pKas, exp_pKa_ranks, pred_pKa_ranks, submission_ID])
    
    
# Convert to pandas dataframe
df_compare_ranks_hungarian = pd.DataFrame(pKa_rank_comparison_list, columns = ['mol ID', 'ordered match', 'pKa IDs', 'pKa (exp)', 
                                               'pKa (pred)', 'pKa rank (exp)', 'pKa rank (pred)', 'submission ID'])

# Print out pKas that don't preserve increasing order when matched my Hungarian algorithm
df_unordered_matching_hungarian = df_compare_ranks_hungarian[df_compare_ranks_hungarian["ordered match"] == False]
df_unordered_matching_hungarian

Empty DataFrame
Columns: [mol ID, ordered match, pKa IDs, pKa (exp), pKa (pred), pKa rank (exp), pKa rank (pred), submission ID]
Index: []

In [97]:
# Just SM18
df_unordered_matching_hungarian_SM18 = df_compare_ranks_hungarian[df_compare_ranks_hungarian["mol ID"] == 'SM18']

# SM18 prediction of yqkga method 
df_unordered_matching_hungarian_SM18_yqkga = df_unordered_matching_hungarian_SM18[df_compare_ranks_hungarian["submission ID"] == 'yqkga']
df_unordered_matching_hungarian_SM18_yqkga

/Users/isikm/opt/anaconda/envs/sampl6_pKa/lib/python3.6/site-packages/ipykernel_launcher.py:5: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """


mol ID  ordered match                            pKa IDs  \
601   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
602   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   
603   SM18           True  [SM18_pKa1, SM18_pKa2, SM18_pKa3]   

               pKa (exp)          pKa (pred)   pKa rank (exp)  \
601  [2.15, 9.58, 11.02]  [2.14, 7.53, 9.26]  [1.0, 2.0, 3.0]   
602  [2.15, 9.58, 11.02]  [2.14, 7.53, 9.26]  [1.0, 2.0, 3.0]   
603  [2.15, 9.58, 11.02]  [2.14, 7.53, 9.26]  [1.0, 2.0, 3.0]   

     pKa rank (pred) submission ID  
601  [1.0, 2.0, 3.0]         yqkga  
602  [1.0, 2.0, 3.0]         yqkga  
603  [1.0, 2.0, 3.0]         yqkga

Hungarian matching algorithm doesn't always make matches that break the natural order of pKa values.
Only in cases where the order preserving match and the unordered match have the same cost value, then the results are random.

The hungarian collection set of branch `pKa_typeIII_analysis5` (commit b1bef28) doesn't have any unordered matches.  
The hungarian collection set of branch `pKa_typeIII_analysis3` (commit 70d828e) has unordered matched for SM18 pKas for the following submission files:
0hxtm, yqkga, ryzue, yc70m
